NorthstarDLL/mods/autodownload/moddownloader.cpp

```diff
@@ -0,0 +1,525 @@
+	{
+		return false;
+	}
+
+	std::unordered_map<std::string, VerifiedModVersion> versions = verifiedMods[modName.data()].versions;
+	return versions.count(modVersion.data()) != 0;
+}
+
+void ModDownloader::ExtractMod(fs::path modPath)
+{
+	unzFile file;
+	std::string name;
+	fs::path modDirectory;
+
+	file = unzOpen(modPath.generic_string().c_str());
+	if (file == NULL)
+	{
+		spdlog::error("Cannot open archive located at {}.", modPath.generic_string());
+		goto EXTRACTION_CLEANUP;
+	}
+
+	unz_global_info64 gi;
+	int status;
+	status = unzGetGlobalInfo64(file, &gi);
+	if (file != UNZ_OK)
+	{
+		spdlog::error("Failed getting information from archive (error code: {})", status);
+	}
+
+	// Mod directory name (removing the ".zip" fom the archive name)
+	name = modPath.filename().string();
+	name = name.substr(0, name.length() - 4);
+	modDirectory = GetRemoteModFolderPath() / name;
+
+	for (int i = 0; i < gi.number_entry; i++)
+	{
+		char zipFilename[256];
+		unz_file_info64 fileInfo;
+		status = unzGetCurrentFileInfo64(file, &fileInfo, zipFilename, sizeof(zipFilename), NULL, 0, NULL, 0);
+
+		// Extract file
+		{
+			std::error_code ec;
+			fs::path fileDestination = modDirectory / zipFilename;
+			spdlog::info("=> {}", fileDestination.generic_string());
+
+			// Create parent directory if needed
+			if (!std::filesystem::exists(fileDestination.parent_path()))
+			{
+				spdlog::info("Parent directory does not exist, creating it.", fileDestination.generic_string());
+				if (!std::filesystem::create_directories(fileDestination.parent_path(), ec) && ec.value() != 0)
+				{
+					spdlog::error("Parent directory ({}) creation failed.", fileDestination.parent_path().generic_string());
+					goto EXTRACTION_CLEANUP;
+				}
+			}
+
+			// If current file is a directory, create directory...
+			if (fileDestination.generic_string().back() == '/')
+			{
+				// Create directory
+				if (!std::filesystem::create_directory(fileDestination, ec) && ec.value() != 0)
+				{
+					spdlog::error("Directory creation failed: {}", ec.message());
+					goto EXTRACTION_CLEANUP;
+				}
+			}
+
+			// ...else create file
+			else
+			{
+				// Ensure file is in zip archive
+				if (unzLocateFile(file, zipFilename, 0) != UNZ_OK)
+				{
+					spdlog::error("File \"{}\" was not found in archive.", zipFilename);
+					goto EXTRACTION_CLEANUP;
+				}
+
+				// Create file
+				int bufferSize;
+				void* buffer;
+				int err = UNZ_OK;
+				FILE* fout = NULL;
+
+				// Open zip file to prepare its extraction
+				status = unzOpenCurrentFile(file);
+				if (status != UNZ_OK)
+				{
+					spdlog::error("Could not open file {} from archive.", zipFilename);
+					goto EXTRACTION_CLEANUP;
+				}
+
+				// Create destination file
+				fout = fopen(fileDestination.generic_string().c_str(), "wb");
+				if (fout == NULL)
+				{
+					spdlog::error("Failed creating destination file.");
+					goto EXTRACTION_CLEANUP;
+				}
+
+				// Allocate memory for buffer
+				bufferSize = 8192;
```

Comment:
Would prefer if this was initialised where it is declared (or vice versa)
Also bufferSize could probably be const

Reply:
Addressed in 99127c1833cd36d30187f77eabdc2bb0e8dc891a and 915583d88e2f367aa46c5591f7e31f19e4845ba3